common/CommonAPI.cpp

```diff
@@ -0,0 +1,1030 @@
+		{
+			nbl::video::IGPUImageView::SCreationParams view_params;
+			view_params.format = img->getCreationParameters().format;
+			view_params.viewType = asset::IImageView<nbl::video::IGPUImage>::ET_2D;
+			view_params.subresourceRange.aspectMask = asset::IImage::EAF_COLOR_BIT;
+			view_params.subresourceRange.baseMipLevel = 0u;
+			view_params.subresourceRange.levelCount = 1u;
+			view_params.subresourceRange.baseArrayLayer = 0u;
+			view_params.subresourceRange.layerCount = 1u;
+			view_params.image = std::move(img);
+
+			view[0] = device->createImageView(std::move(view_params));
+			assert(view[0]);
+		}
+
+		if (useDepth) {
+			nbl::video::IGPUImage::SCreationParams imgParams;
+			imgParams.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0u);
+			imgParams.type = asset::IImage::ET_2D;
+			imgParams.format = depthFormat;
+			imgParams.extent = { width, height, 1 };
+			imgParams.usage = asset::IImage::E_USAGE_FLAGS::EUF_DEPTH_STENCIL_ATTACHMENT_BIT;
+			imgParams.mipLevels = 1u;
+			imgParams.arrayLayers = 1u;
+			imgParams.samples = asset::IImage::ESCF_1_BIT;
+
+			auto depthImg = device->createImage(std::move(imgParams));
+			auto depthImgMemReqs = depthImg->getMemoryReqs();
+			depthImgMemReqs.memoryTypeBits &= device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			auto depthImgMem = device->allocate(depthImgMemReqs, depthImg.get());
+
+			nbl::video::IGPUImageView::SCreationParams view_params;
+			view_params.format = depthFormat;
+			view_params.viewType = asset::IImageView<nbl::video::IGPUImage>::ET_2D;
+			view_params.subresourceRange.aspectMask = asset::IImage::EAF_DEPTH_BIT;
+			view_params.subresourceRange.baseMipLevel = 0u;
+			view_params.subresourceRange.levelCount = 1u;
+			view_params.subresourceRange.baseArrayLayer = 0u;
+			view_params.subresourceRange.layerCount = 1u;
+			view_params.image = std::move(depthImg);
+
+			view[1] = device->createImageView(std::move(view_params));
+			assert(view[1]);
+		}
+
+		nbl::video::IGPUFramebuffer::SCreationParams fb_params;
+		fb_params.width = width;
+		fb_params.height = height;
+		fb_params.layers = 1u;
+		fb_params.renderpass = renderpass;
+		fb_params.flags = static_cast<nbl::video::IGPUFramebuffer::E_CREATE_FLAGS>(0);
+		fb_params.attachmentCount = (useDepth) ? 2u : 1u;
+		fb_params.attachments = view;
+
+		fbo->begin()[i] = device->createFramebuffer(std::move(fb_params));
+		assert(fbo->begin()[i]);
+	}
+	return fbo;
+}
+
+bool CommonAPI::createSwapchain(
+	const nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>&& device,
+	nbl::video::ISwapchain::SCreationParams& params,
+	uint32_t width, uint32_t height,
+	// nullptr for initial creation, old swapchain for eventual resizes
+	nbl::core::smart_refctd_ptr<nbl::video::ISwapchain>& swapchain
+)
+{
+	auto oldSwapchain = swapchain;
+
+	nbl::video::ISwapchain::SCreationParams paramsCp = params;
+	paramsCp.width = width;
+	paramsCp.height = height;
+	paramsCp.oldSwapchain = oldSwapchain;
```

Comment:
1. Why would the user provide the params when he can just set these in the params?
seems redundant.

The only use for this function I see is hiding the API things.
2. I also don't understand why we removed that from ILogicalDevice and now don't have an API agnostic `createSwapchain` function in Nabla, Can you clarify?

Reply:
1. That's more of a convenience thing, the examples don't really touch the parameters outside of moving them around

2. That was done for decoupling swapchain from logical device, do you want there to be a function that creates it?

Reply:
> That was done for decoupling swapchain from logical device
Ok but can you remind me why we want them decoupled?